### PR TITLE
Create basic test suite

### PR DESCRIPTION
**Pull request by shssoichiro**
This does not cover every function or behavior in the package, but does create a test suite that can ensure basic sanity and be expanded upon in the future.

### CODE REVIEW

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -25,7 +25,13 @@ jobs:
...
       - name: Running flake8
-        run: flake8 vstools
+        run: |
+          flake8 vstools
+          flake8 tests
       - name: Running mypy
-        if: steps.dependencies.outcome == 'success'
-        run: mypy vstools
\ No newline at end of file
+        continue-on-error: true
```
Comment (shssoichiro) [inline .github/workflows/lint.yml:32]: Seems there's a lot of stuff in the codebase that mypy is yelling about... not sure if it necessarily makes sense to fix it all in this PR, it could get quite large.

File: tests/enums/test_color.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+        result = Matrix.from_res(clip)
+        self.assertEqual(result, Matrix.BT470BG)
+
+    def test_from_video_rgb(self) -> None:
+        clip = vs.core.std.BlankClip(format=vs.RGB24)
+        result = Matrix.from_video(clip)
+        self.assertEqual(result, Matrix.RGB)
+
+    def test_from_video_uhd(self) -> None:
+        clip = vs.core.std.BlankClip(format=vs.YUV420P8, width=3840, height=2160)
```
Comment (Setsugennoao) [inline tests/enums/test_color.py:54]: the from_video tests should check for the property

File: .github/workflows/lint.yml (hunk 1 above)
Comment (Setsugennoao) [reply .github/workflows/lint.yml:32]: then ignore and revert the changes

File: vstools/enums/color.py [hunk 3]
```diff
@@ -988,6 +988,13 @@ def _missing_(cls: type[ColorRange], value: Any) -> ColorRange | None:
     def from_res(cls, frame: vs.VideoNode | vs.VideoFrame) -> ColorRange:
         """Guess the color range from the frame resolution."""
 
+        from ..utils import get_var_infos
+
+        fmt, w, h = get_var_infos(frame)
```
Comment (Setsugennoao) [inline vstools/enums/color.py:993]: can the fixes be moved to another PR? like a generic bugfix/typos one

File: vstools/enums/generic.py [hunk 4]
```diff
@@ -86,6 +86,7 @@ def get_offsets(
         if isinstance(chroma_loc, vs.VideoNode):
             assert chroma_loc.format  # type: ignore
             subsampling = (chroma_loc.format.subsampling_w, chroma_loc.format.subsampling_h)  # type: ignore
+            offsets = (0, 0)
```
Comment (Setsugennoao) [inline vstools/enums/generic.py:89]: please move this in the else at the end of the block, L99

File: vstools/functions/normalize.py [hunk 5]
```diff
@@ -237,7 +237,7 @@ def normalize_ranges(clip: vs.VideoNode, ranges: FrameRangeN | FrameRangesN) ->
             if start is None:
                 start = 0
             if end is None:
-                end = clip.num_frames - 1
```
Comment (Setsugennoao) [inline vstools/functions/normalize.py:240]: this is not a bug, please revert

File: tests/functions/test_normalize.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+class TestNormalize(TestCase):
+    def test_flatten(self) -> None:
+        result = flatten(["a", "b", ["c", "d", ["e"]]])
+        self.assertEqual(list(result), ["a", "b", "c", "d", "e"])
+
+    def test_normalize_ranges(self) -> None:
+        clip = vs.core.std.BlankClip(length=1000)
+
+        self.assertEqual(normalize_ranges(clip, (None, None)), [(0, 1000)])
```
Comment (Setsugennoao) [inline tests/functions/test_normalize.py:16]: the last inclusive range should be 999, not 1000

File: vstools/functions/normalize.py [hunk 7]
```diff
@@ -246,7 +246,7 @@ def normalize_ranges(clip: vs.VideoNode, ranges: FrameRangeN | FrameRangesN) ->
             start = clip.num_frames - 1 + start
 
         if end < 0:
-            end = clip.num_frames - 1 + end
```
Comment (Setsugennoao) [inline vstools/functions/normalize.py:249]: the negative offset should be inclusive too

File: vstools/functions/normalize.py (hunk 5 above)
Comment (shssoichiro) [reply vstools/functions/normalize.py:240]: The examples in this function are incorrect then. I can revert this change and update the examples instead. (The tests are identical to the examples.)

File: tests/enums/test_color.py (hunk 2 above)
Comment (shssoichiro) [reply tests/enums/test_color.py:54]: Added more tests around this

File: vstools/enums/color.py (hunk 3 above)
Comment (shssoichiro) [reply vstools/enums/color.py:993]: Some of the tests fail without the changes. I suppose I can temporarily @ignore those and unignore in the fix PR.

File: .github/workflows/lint.yml (hunk 1 above)
Comment (shssoichiro) [reply .github/workflows/lint.yml:32]: With `continue-on-error: true` seems to ignore it and go on to the tests. Which seems like they're failing because of some missing pip dependencies, I'll have to get those fixed up.

File: tests/enums/test_color.py [hunk 8]
```diff
@@ -50,6 +51,12 @@ def test_from_video_rgb(self) -> None:
         result = Matrix.from_video(clip)
         self.assertEqual(result, Matrix.RGB)
 
+    def test_from_video_property(self) -> None:
+        clip = vs.core.std.BlankClip(format=vs.YUV420P8, width=3840, height=2160)
+        clip = vs.core.std.SetFrameProp(clip, "_Matrix", Matrix.BT709)
```
Comment (Setsugennoao) [inline tests/enums/test_color.py:56]: Do you think we should mmake two different tests? One with SetFrameProp and the other with Matrix.BT709.apply 🤔

**Comment by shssoichiro** [comment]
@Setsugennoao anything still needed on this?

**Comment by Setsugennoao** [comment]
Yes, I need to check some test.

But also, I wanted to expand this PR to cover all the files, and start adding more internal stuff myself.

**Comment by shssoichiro** [comment]
I think it would be better to merge this (with any fixes/checks that are needed) rather than waiting for all functions to have tests. More tests can always be added later, but leaving this open for a potentially long period of time means these tests are not serving their purpose in the meantime, and potentially will encounter annoying rebase issues. (Pretty sure at least one test needs to be updated as a result of #79)

**Comment by Setsugennoao** [comment]
The thing is I have to check a lot of values, because some don't check out.
I was looking into it and boy, I don't feel like expanding it or maintaining it, really.

Btw, can you make it a separate GitHub action?

**Comment by shssoichiro** [comment]
I hate writing tests as much as the next dev, but it's important to keep stuff from breaking. I'm willing to help as much as I'm able, although I simply don't know all of the internals (especially for the whole of IEW).

And yeah I'll move it to a separate action, np.
